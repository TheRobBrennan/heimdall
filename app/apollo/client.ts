import fetch from "cross-fetch"
import { useMemo } from "react"

import { ApolloClient, createHttpLink, InMemoryCache } from "@apollo/client"
import { setContext } from "@apollo/client/link/context"

let apolloClient

const httpLink = createHttpLink({
  uri: `api/graphql`,
  fetch,
})

/* istanbul ignore next */
const authLink = setContext(async () => {
  // See https://www.apollographql.com/docs/react/networking/authentication/#header for more details
  // Get our session detail from the back-end API
  let authorization = ""
  try {
    const data = await (await fetch("api/session")).json()
    const { accessToken } = data
    authorization = `Bearer ${accessToken}`
  } catch (e) {
    // No session data available; no authorization header will be sent
  }

  // Return the headers to the context so httpLink can read them
  return {
    headers: {
      authorization,
    },
  }
})

function createApolloClient(initialState) {
  const ssrMode = typeof window === "undefined"

  return new ApolloClient({
    ssrMode,
    link: authLink.concat(httpLink),
    cache: new InMemoryCache().restore(initialState),
  })
}

export function initializeApollo(initialState = null) {
  const _apolloClient = apolloClient ?? createApolloClient(initialState)

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // get hydrated here
  if (initialState) {
    _apolloClient.cache.restore(initialState)
  }
  // For SSG and SSR always create a new Apollo Client
  /* istanbul ignore next */
  if (typeof window === "undefined") return _apolloClient
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient

  return _apolloClient
}

/* istanbul ignore next */
export function useApollo(initialState) {
  const store = useMemo(() => initializeApollo(initialState), [initialState])
  return store
}
