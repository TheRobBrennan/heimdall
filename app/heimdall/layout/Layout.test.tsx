import * as React from "react"
import { cleanup, render } from "@testing-library/react"
import { MockedProvider } from "@apollo/client/testing"

import Component from "./Layout"

describe("Our Layout component", () => {
  afterEach(cleanup)

  it("should render without children", () => {
    const { asFragment } = render(
      <MockedProvider mocks={[]} addTypename={false}>
        <Component />
      </MockedProvider>
    )

    expect(asFragment).toMatchSnapshot()
  })

  it("should render with children", () => {
    const { asFragment } = render(
      <MockedProvider mocks={[]} addTypename={false}>
        <Component>
          <h1>An example heading</h1>
          <p>Some content.</p>
        </Component>
      </MockedProvider>
    )

    expect(asFragment).toMatchSnapshot()
  })
})
