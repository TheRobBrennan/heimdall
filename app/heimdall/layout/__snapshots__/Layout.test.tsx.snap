// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Our Layout component should render with children 1`] = `[Function]`;

exports[`Our Layout component should render without children 1`] = `[Function]`;
